File: game/scripts/vscripts/modifiers/upgrades/items/modifier_upgrade_item_basher.lua
modifier_upgrade_item_basher = class({})
modifier_upgrade_item_basher_attack = class({})

function modifier_upgrade_item_basher:RemoveOnDeath() return false end
function modifier_upgrade_item_basher:IsPurgable() return false end

function modifier_upgrade_item_basher:OnCreated(params)
    self.damage_per_stack = 10
    self.extra_health = 100
    
    if IsServer() then
        self.proc_chance_pct =  25
    end
end

function modifier_upgrade_item_basher:DeclareFunctions()
    return { 
		MODIFIER_PROPERTY_HEALTH_BONUS,
        MODIFIER_PROPERTY_PREATTACK_BONUS_DAMAGE,
    }
end

function modifier_upgrade_item_basher:GetModifierPreAttack_BonusDamage(params)
    return self.damage_per_stack
end

function modifier_upgrade_item_basher:GetModifierHealthBonus()
	return self.extra_health
end

function modifier_upgrade_item_basher:OnEvent(params)
    if params.iEventId == MODIFIER_EVENTS.ON_BASIC_ATTACK_STARTED then
        self:GetParent():AddNewModifier(self:GetParent(), nil, 'modifier_upgrade_item_basher_attack', {
            proc_chance_pct = self.proc_chance_pct
        })
    end
end

function modifier_upgrade_item_basher_attack:IsHidden()
    return true
end

function modifier_upgrade_item_basher_attack:OnCreated(params)
    if IsServer() then
        self.proc_chance_pct = params.proc_chance_pct
        self.damage_table = {
            attacker = self:GetParent(),
            damage = 10,
            damage_type = DAMAGE_TYPE_MAGICAL,
        }
        self.stun_duration = 1.0
    end
end

function modifier_upgrade_item_basher_attack:OnEvent(params)
    if params.iEventId == MODIFIER_EVENTS.ON_BASIC_ATTACK_ENDED then
        self:Destroy()
    end
    if params.iEventId == MODIFIER_EVENTS.ON_BASIC_ATTACK_LANDED then
        if IsServer() then
            if RandomInt(1, 100) <= self.proc_chance_pct then
                if params.hTarget:IsAlive() then
                    self:Stun(self:GetParent(), params.hTarget)
                end
            end
        end
    end
end

function modifier_upgrade_item_basher_attack:Stun(hSource, hTarget)
    local direction = Direction2D(hSource:GetAbsOrigin(), hTarget:GetAbsOrigin())
    EFX("particles/units/heroes/hero_spirit_breaker/spirit_breaker_greater_bash.vpcf", PATTACH_CUSTOMORIGIN, hTarget, {
        cp0 = {
            ent = hTarget,
            point = 'attach_hitloc'
        },
        cp0f = direction,
        release = true,
    })

    self.damage_table.victim = hTarget
    ApplyDamage(self.damage_table)
    hTarget:AddNewModifier(hSource, nil, "modifier_generic_stunned", { duration = self.stun_duration })

    EmitSoundOn("DOTA_Item.SkullBasher", hTarget)
end

function modifier_upgrade_item_basher:GetTexture()
    return 'item_basher'
end

if IsClient() then require("wrappers/modifiers") end
Modifiers.OnEvent(modifier_upgrade_item_basher)
Modifiers.OnEvent(modifier_upgrade_item_basher_attack)

File: src/vscripts/modifiers/upgrades/items/modifier_upgrade_item_basher.ts
import { registerModifier } from "../../../lib/dota_ts_adapter";
import { direction2D } from "../../../util";
import { ModifierCombatEvents } from "../../modifier_combat_events";
import { ModifierStun } from "../../modifier_stunned";
import { ModifierUpgrade } from "../modifier_upgrade";

@registerModifier("modifier_upgrade_item_basher")
class ModifierItemBasher extends ModifierUpgrade {
    damagePerStack = 10;
    extraHealth = 100;
    procChancePct = 25;

    DeclareFunctions() {
        return [ModifierFunction.HEALTH_BONUS, ModifierFunction.PREATTACK_BONUS_DAMAGE];
    }

    GetModifierPreAttack_BonusDamage() {
        return this.damagePerStack;
    }

    GetModifierHealthBonus() {
        return this.extraHealth;
    }

    OnBasicAttackStarted() {
        if (RandomInt(1, 100) <= this.procChancePct) {
            ModifierItemBasherAttack.apply(this.parent, this.parent, undefined, {});
        }
    }
}

@registerModifier("modifier_upgrade_item_basher_attack")
class ModifierItemBasherAttack extends ModifierCombatEvents<undefined> {
    stunDuration = 1.0;
    damage = 10;

    IsHidden() {
        return true;
    }

    OnBasicAttackEnded() {
        this.Destroy();
    }

    OnBasicAttackLanded(event: { target: CDOTA_BaseNPC }): void {
        if (IsServer() && event.target.IsAlive()) {
            this.Stun(this.GetParent(), event.target);
        }
    }

    Stun(source: CDOTA_BaseNPC, target: CDOTA_BaseNPC) {
        const direction = direction2D(source.GetAbsOrigin(), target.GetAbsOrigin());
        EFX("particles/units/heroes/hero_spirit_breaker/spirit_breaker_greater_bash.vpcf", ParticleAttachment.CUSTOMORIGIN, target, {
            cp0: {
                ent: target,
                point: AttachLocation.hitloc
            },
            cp0f: direction,
            release: true
        });

        ApplyDamage({
            victim: target,
            attacker: this.parent,
            damage: this.damage,
            damage_type: DamageTypes.MAGICAL
        });

        ModifierStun.apply(target, source, undefined, { duration: this.stunDuration });
        EmitSoundOn("DOTA_Item.SkullBasher", target);
    }

    GetTexture() {
        return "item_basher";
    }
}
